src/NUnitFramework/framework/Internal/Commands/TimeoutCommand.cs

```diff
@@ -0,0 +1,83 @@
+﻿// ***********************************************************************
+// Copyright (c) 2017 Charlie Poole
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace NUnit.Framework.Internal.Commands
+{
+    using Execution;
+    using Interfaces;
+
+    /// <summary>
+    /// SetUpTearDownCommand runs SetUp methods for a suite,
+    /// runs the test and then runs TearDown methods.
+    /// </summary>
+    public class TimeoutCommand : BeforeAndAfterTestCommand
+    {
+        Timer _commandTimer;
+        private bool _commandTimedOut = false;
+        
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SetUpTearDownCommand"/> class.
+        /// </summary>
+        /// <param name="innerCommand">The inner command</param>
+        /// <param name="timeout">Timeout value</param>
+        public TimeoutCommand(TestCommand innerCommand, int timeout)
+            : base(innerCommand)
+        {
+            Guard.ArgumentValid(innerCommand.Test is TestMethod, "SetUpTearDownCommand may only apply to a TestMethod", "innerCommand");
+            Guard.ArgumentValid(timeout > 0, "Timeout value must be greater than zero", "timeout");
+
+            BeforeTest = (context) =>
+            {
+                var testThread = Thread.CurrentThread;
+
+                // Create a timer to cancel the current thread
+                _commandTimer = new Timer(
+                    (o) =>
+                    {
+                        _commandTimedOut = true;
+                        testThread.Abort();
+                        // No join here, since the thread doesn't really terminate
```

Comment:
Oh wait - probably not - I'd not yet got to the bit where this was caught later on!